### ISSUE TITLE
URIs of individual terms should be dereferenceable over HTTP

### ISSUE THREAD

**Issue by rjyounes** [semanticarts/gist#130]
All HTTP URIs in the Linked Data universe should be deferenceable. This is a critical part of Tim Berners-Lee's four Linked Data principles:

2. Use HTTP URIs, so that people can look up those names.
3. When someone looks up a URI, provide useful information...

Currently,  the gist ontology URIs are dereferenceable, but the URIs of individual terms are not. This appears to be a violation of the principles stated above, as I understand them. In many implementations I am familiar with, the URIs of individual terms can be dereferenced to receive just that individual term's definition (when the namespace ends in a slash) or the entire ontology (when the namespace ends in a hash). Examples: BIBFRAME, Schema.

This in turn suggests to me that the URIs of individual terms should be versioned, for the same reason that owl:imports should refer to versioned URIs: otherwise, the user cannot control what is returned from his request;  see discussion on issue #127. 

In summary, the proposal is that gist terms should have URIs something like https://ontologies.semanticarts.com/gist9.0.0/Person; that these should be dereferenceable; that, like ontologies, the unversioned URI should redirect to the latest version so that a user can control what is returned from his request; and that https://ontologies.semanticarts.com/gist9.0.0/Person and http://ontologies.semanticarts.com/gist8.0.0/Person are independently dereferenceable.

This of course returns us to my earlier discussion on versioning and namespaces, where my proposal to version term URIs was countered by the argument that URIs should be stable and not change every time a new version is released. This is a valid objection, but I'd like to discuss how these two principles should be reconciled. 

Even if the gist namespace were hash- rather than slash-final, a request for an individual term should at least return the entire ontology rather than a 404 error. However, I believe the namespace was recently changed from hash to slash, presumably for good reasons, so that is likely not a realistic option.

At a minimum, this issue can be partially addressing by making unversioned term URIs dereferenceable.

**Comment by uscholdm** [comment]
URIs of concepts used to be dereferenceable.  IT returned a text page for humans; I don't think it ever returned OWL to a machine.  Not sure what happened to this functionality.

**Comment by uscholdm** [comment]
Versioning URIs of individual terms is a bit of a holy grail. Opens up major can of dependency worms, for which no rubust technology exists to support.

**Comment by marksem** [comment]
I think when you look up a term URI, which is unversioned, you get the current released version's term.   Not sure how else you'd do it.

**Comment by marksem** [comment]
I think I could write a little API/service that we could run on our web site that would do this.

**Comment by Jamie-SA** [comment]
There is a plan to serve SA ontologies and terms from the DCA server with a lot of additional functionality. I am starting to work on the URI resolver which will handle mapping various URI prefixes to content in triplestores. This will eventually provide content negotiation and additional functionality for exploring ontologies.

**Comment by rjyounes** [comment]
Ignore issue of versioned IRIs (which has been dropped) and focus on dereferenceability. Adding to current project to get a status update from @Jamie-SA.

**Comment by stevenchalem** [comment]
@rjyounes Is there a standard or common way of dereferencing a term? (What info to include, how to present it.) Or, do you have an example you like?

**Comment by rjyounes** [comment]
Via content negotiation, it should return either the RDF or, if HTML is requested, some sort of documentation, such as Widoco. I don't know if this is a standard but this is what I see online. 

Sometimes RDF is returned in the browser, sometimes in a download file. I have a preference for the former - you can always paste it into a file if that's what you want - because if you just want to view it quickly you want an immediate response.

IMO it would be ideal to return just the portion of the ontology or documentation for that individual term, but I don't see it done that way in most of the major the examples I've looked at - skos, prov, dcterms. Schema does (at least in the HTML version). Most are returning the entire document, but in the case of HTML the requested term is displayed at the top of the page. Perhaps there are technical barriers to returning just the single term - @Jamie-SA could comment.

**Comment by rjyounes** [comment]
Section 2.3 of the book _Linked Data_ by Heath and Bizer provides an excellent discussion. I don't think there's a free version online, though.

**Comment by stevenchalem** [comment]
Thanks. I found some good info here:
http://tomheath.com/papers/bizer-heath-berners-lee-ijswis-linked-data.pdf

**Comment by Jamie-SA** [comment]
There is a bit of a technical problem if the namespace has a hash symbol (`#`) in it, which many old namespaces have. The problem is that what follows the hash is not sent to the server, therefore the server can only see that namespace being requested and not the specific term.

This can be addressed with javascript on the client side, but that requires programming versus just serving a file.

**Comment by rjyounes** [comment]
@Jamie-SA Right, but fortunately our namespace ends in slash rather than hash.

**Comment by Jamie-SA** [comment]
And this is one of the reasons it was changed years ago (it used to end with a hash).

**Comment by uscholdm** [comment]
@rjyounes @Jamie-SA 
Jamie, can you have another look at this, it is deferred, should we close or activate?

**Comment by rjyounes** [comment]
This needs to be done before gist can be added to [Linked Open Vocabularies (LOV)](https://lov.linkeddata.es/).
